src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -544,32 +572,47 @@ public TranslogStats stats() {
         }
     }
 
-    private boolean isReferencedTranslogId(long translogId) {
-        return translogId >= lastCommittedTranslogId;
+    private boolean isReferencedGeneration(long generation) { // used to make decisions if a file can be deleted
+        return generation >= lastCommittedTranslogFileGeneration;
     }
 
-    private final class InternalChannelReference extends ChannelReference {
-        final long translogId;
+    public TranslogConfig getConfig() {
+        return config;
+    }
 
-        public InternalChannelReference(long translogId, Path file, OpenOption... openOptions) throws IOException {
-            super(file, openOptions);
-            this.translogId = translogId;
-        }
 
+    private final class OnCloseRunnable implements Callback<ChannelReference> {
         @Override
-        protected void closeInternal() {
-            super.closeInternal();
+        public void handle(ChannelReference channelReference) {
             try (ReleasableLock lock = writeLock.acquire()) {
-                if (isReferencedTranslogId(translogId) == false) {
-                    // if the given path is not the current we can safely delete the file since all references are released
-                    logger.trace("delete translog file - not referenced and not current anymore {}", file());
-                    IOUtils.deleteFilesIgnoringExceptions(file());
+                if (isReferencedGeneration(channelReference.getGeneration()) == false) {
+                    Path translogPath = channelReference.getPath();
+                    assert channelReference.getPath().getParent().equals(location) : "translog files must be in the location folder: " + location + " but was: " + translogPath;
+                    // if the given translogPath is not the current we can safely delete the file since all references are released
+                    logger.trace("delete translog file - not referenced and not current anymore {}", translogPath);
+                    IOUtils.deleteFilesIgnoringExceptions(translogPath);
+                    IOUtils.deleteFilesIgnoringExceptions(translogPath.resolveSibling(getCommitFileName(channelReference.getGeneration())));
+
+                }
+                try (DirectoryStream<Path> stream = Files.newDirectoryStream(location)) {
```

Comment:
Hmm we do a new directory listing of the translog directory every time a TranslogReader or its clone is closed ... but is this really necessary, or purely defensive?  I.e., even with crashing, we should always eventually see onClose called on a translog that's truly unreferenced?  Hmm, except for that "forceNewTranslog=true" case maybe?